server/api/engines/supplement.js

```diff
@@ -29,6 +34,86 @@ class SupplementEngine
 
     //------------------------------------------------------------------------------------------------------------------
 
+    async $validateSupplementPath(data, suppDef, systemPrefix, account)
+    {
+        // FIXME: This is a circular dependency, and I hate it.
+        const suppMan = require('../managers/supplement');
```

Comment:
Since I have no idea what this function is supposed to do, I can't really give an decent advice.

If it's to ensure a supplement is submitted to the correct http path, we could include that as part of an entities structure, and make that the id.

Reply:
If you look a few lines down, we use it:

```javascript
const filtered = (await suppMan.filterSupplementsByPermissions(ids, suppDef.type, systemPrefix, account))
                .map((obj) => obj.id);
```

What this does, is batch retrieve ids, and then filter them by the permissions, so you only get back the list you're allowed to look at. We then test every item in `supplement` to see if it's contained in `filtered`, and if it is, keep it.

(The reason for the later is that `supplement` can an object of `id` and `ranks`, which are the ranks of this supplement that've been purchased. Other metadata could, in theory, be stored in this object as well, we need to preserve the entire object if we're not filtering it out.)

Reply:
Basically, in short, the only way to break this would be to refactor the code a bit, which would be doable. There's a chance this is improved in TS, so the circular dep doesn't break everything.